### ISSUE TITLE
Proposal: Add option to format code with leading commas

### ISSUE THREAD

**Issue by SetTrend** [Microsoft/vscode#24869]
Can we get an option in VS Code to format all lists with leading commas instead of trailing commas?

This would so much simplify editing code and it would also greatly simplify the IDE's code completion.

With the suggested option set to `true`, comma separated lists would be formatted like this:

```js
{ "workbench.iconTheme": "vs-seti"
, "editor.lineNumbers": "off"
, "editor.tabSize": 2
, "editor.insertSpaces": false
, "editor.autoClosingBrackets": false
}
```

function calls using argument wrapping would look like this:

```js
call( 1
    , "test"
    , new Console()
    );
```

a nested sample would look like this:

```json
{ "version": "0.2.0"
, "configurations":
  [ { "type": "node"
    , "request": "launch"
    , "name": "Programm starten"
    , "program": "${file}"
    }
  , { "type": "node"
    , "request": "attach"
    , "name": "An den Port anfügen"
    , "address": "localhost"
    , "port": 5858
    }
  ]
}
```

This formatting style would result in much cleaner, tidier code.
<br/>

### Moreover, adding a new property to an object would be a snap:

Either copy/paste from an existing property (=> no more syntax errors due to redundant trailing commas, see screenshot below):

![comma error](https://cloud.githubusercontent.com/assets/9283914/25074723/8e40d278-2302-11e7-94d8-d8b35d3321ce.png)

...  or create a new property and you might immediately get IntelliSense right after typing the leading comma of the new property.

**Comment by maneesht** [comment]
@aeschli I'd like to help out with this, if no one's working on it.

**Comment by SetTrend** [comment]
@aeschli :
Thanks for taking the time! 👍 

Yet, my suggestion is not only dealing with JSON, but with all C-style languages. (See the function call example and the T-SQL example I gave above).

**Comment by aeschli** [comment]
@SetTrend Every language has a different formatter, and very few shared the implementation. IMO sharing formatter settings is not realistic, there are too many subtle differences between languages.

**Comment by aeschli** [comment]
@maneesht The JSON formatter implementation is in https://github.com/Microsoft/vscode-json-languageservice/blob/master/src/services/jsonFormatter.ts. 
You'll also find test cases there. 
Pull requests are highly welcome, please create them in the https://github.com/Microsoft/vscode-json-languageservice repository.

**Comment by SetTrend** [comment]
@aeschli I understand. Thanks for enlightening me.

I'm not sure about the current VS Code implementation, yet (I can only speak for myself here) in every C-style language I'm using I'm always configuring all the common formatting settings alike. I do this in VS Code and in VS Enterprise. (Which is rather tedious, but I wouldn't want to switch formatting styles from language to language - would anyone?)

Wouldn't it make sense to provide a common base formatter for C-style languages that could be used and overriden by language services to add peculiarities to a language's formatting?
<br/>
I believe, the decision to use leading instead of trailing commas is a design decision similar to using _folding_ or preferring _spaces_ to _tabs_, or vice versa. Folding, for instance, is implemented per language, too. Yet, the option in VS code is a global one. That's why I'd like to propose a global setting, like `editor.listFormatting`, using an enumeration, like

```ts
enum ListFormatting
{ TrailingComma
, LeadingComma
, LeadingCommaCompact
}
```

The difference between `LeadingComma` and `LeadingCommaCompact` may be the closing brackets: If consecutive closing brackes appear, they should probably not take a line on their own, depending on user preference.

`LeadingCommaCompact` would, after a comma separated line-split list has been formatted, take the additional step of replacing all whitespace between closing brackets with a space character.

A generic implementation (valid for all languages) would be rather simple, because in all languages a comma separated, line-split list is always characterized by matching brackets containing commas optionally accompanied by one or more line-breaks.
<br/>
### A generic (pseudo) implementation may look like the following

```ts
let lists : List[] = editor.getLists();  // that's a ubiquitios function, fitting lists in all kinds of languages

foreach (var i = 0;i < lists.length;++i)
  lists[i].Format();  // that's a ubiquitios function, too, fitting lists in all kinds of languages

editor.replaceLists(lists);  // that's another ubiquitios function, fitting lists in all kinds of languages
```

```ts
class List
{
  public SelEnd : number;
  public Content: string;

  List(editorText : string, public SelStart: number)
  {
    // find the next beginning of a comma separated, line-split list
    this.SelStart = ...;

    // find the end of that list
    this.SelEnd = ...;

    this.Content = editorText.substring(this.SelStart, this.SelEnd);
  }

  function Format()
  {
    switch (Application.Settings.Editor.listFormatting)
    {
      case ListFormatting.TrailingComma:
        this.Content = ...;
        break;

      case ListFormatting.LeadingComma:
      case ListFormatting.LeadingCommaCompact
        this.Content = ...;
        break;
    }

    if (Application.Settings.Editor.listFormatting === ListFormatting.LeadingCommaCompact)
      this.Content = this.Content.replace(/([\])}])\s+([\])}])/, "$1 $2");
  }
}
```

```ts
Editor.prototype.getLists = function()
{
  let result : List[] = [];

  for (let textPos = 0; textPos < this.TextContent.length;)
  {
    let list : List = new List(this.TextContent, textPos);

    if (list.SelStart < 0) break;   // no more lists found in editor's text buffer?

    result.push(list);
    textPos += list.SelEnd - list.SelStart;
  }

  return result;
}
```

```ts
Editor.prototype.replaceLists = function(lists : List[])
{
  for (let i=lists.length;i>0;)
  {
    let list : List = lists[i];

    this.TextContent = this.TextContent.substr(0, list.SelStart) + list.Content + this.TextContent.substr(list.SelEnd);
  }
}
```

**Comment by maneesht** [comment]
@aeschli How would I test my code for this? Any tips would be appreciated!

**Comment by SetTrend** [comment]
@dbaeumer : May I invite you to this thread? Would you mind having a look at this and share your opinion?

**Comment by phinque** [comment]
I agree with this concept, add formatting for leadingCommas to all languages

**Comment by caub** [comment]
Why do people still use this formatting? objects, arrays, functions allow trailing commas now, it's significantly more readable

**Comment by SetTrend** [comment]
They always did. Same as they always supported leading commas. There is no rule on where to put punctuation characters.

Actually, once you started using leading commas you will immediately recognize the benefit of using it, and you will never want to revert to struggle with the clumsiness of trailing commas.

**Comment by SetTrend** [comment]
They always did. Same as they always supported leading commas. There is no rule on where to put punctuation characters.

Actually, once you started using leading commas you will immediately recognize the benefit of using it, and you will never want to revert to struggle with the clumsiness of trailing commas.

**Comment by SetTrend** [comment]
Just for the sake of showing that this kind of formatting is not an unpopular one:

https://docs.npmjs.com/misc/coding-style#comma-first

**Comment by Vildan** [comment]
For now using Better Align extention. Works pretty well. https://marketplace.visualstudio.com/items?itemName=wwm.better-align
